### PR TITLE
`derive(IntoBytes)`: support `repr(C)` structs with explicit trailing slices

### PR DESCRIPTION
**Pull request by jswrenn**
For example:

    #[repr(C)]
    struct Example {
        leading: u8,
        trailing: [Trailing]
    }

Makes progress towards #1112

### CODE REVIEW

File: src/layout.rs [hunk 1]
```diff
@@ -202,10 +203,71 @@ impl DstLayout {
         }
     }
 
+    /// Constructs a complete `DstLayout` reflecting a `repr(C)` struct with the
+    /// given alignment modifiers and fields.
+    ///
+    /// The accompanying boolean is `true` if the resulting composition of
+    /// fields necessitated static (as opposed to dynamic) padding; otherwise
+    /// false.
```
Comment (joshlf) [inline src/layout.rs]: Can we make this an enum so its meaning is clearer at the call site?

File: src/layout.rs [hunk 2]
```diff
@@ -202,10 +203,71 @@ impl DstLayout {
...
+    ///
+    /// This method cannot be used to match the layout of a record with the
+    /// default representation, as that representation is mostly unspecified.
+    ///
+    /// # Safety
+    ///
+    /// For any definition `repr(C)` struct, if this method is invoked with
```
Comment (joshlf) [inline src/layout.rs]:
```
definition `repr(C)` struct
```

Typo?

File: src/layout.rs [hunk 3]
```diff
@@ -202,10 +203,71 @@ impl DstLayout {
...
+    }
+
     /// Like `Layout::extend`, this creates a layout that describes a record
     /// whose layout consists of `self` followed by `next` that includes the
     /// necessary inter-field padding, but not any trailing padding.
     ///
+    /// The accompanying boolean is `true` if the resulting composition of
+    /// fields necessitated static (as opposed to dynamic) padding; otherwise
+    /// false.
+    ///
```
Comment (joshlf) [inline src/layout.rs]: Same here

File: src/layout.rs [hunk 4]
```diff
@@ -295,59 +357,66 @@ impl DstLayout {
...
     }
 
     /// Like `Layout::pad_to_align`, this routine rounds the size of this layout
     /// up to the nearest multiple of this type's alignment or `repr_packed`
     /// (whichever is less). This method leaves DST layouts unchanged, since the
     /// trailing padding of DSTs is computed at runtime.
     ///
+    /// The accompanying boolean is `true` if the resulting composition of
+    /// fields necessitated static (as opposed to dynamic) padding; otherwise
+    /// false.
```
Comment (joshlf) [inline src/layout.rs]: Same here

File: src/layout.rs [hunk 5]
```diff
@@ -844,7 +931,7 @@ mod tests {
                     core::iter::once(None).chain((0..29).map(|p| NonZeroUsize::new(2usize.pow(p))));
 
                 for pack in packs {
-                    let composite = base.extend(trailing_field, pack);
+                    let composite = base.extend(trailing_field, pack).1;
```
Comment (joshlf) [inline src/layout.rs]: To make this a bit more self-documenting:
```suggestion
                    let (_padding, composite) = base.extend(trailing_field, pack);
```

Same below in various places that `.1` is used.

File: src/layout.rs [hunk 6]
```diff
@@ -1703,6 +1790,44 @@ mod proofs {
...
+    #[kani::proof]
+    fn prove_requires_dynamic_padding() {
+        let layout: DstLayout = kani::any();
+
+        let SizeInfo::SliceDst(size_info) = layout.size_info else {
+            kani::assume(false);
+            loop {}
+        };
+
+        //kani::assume(!layout.requires_dynamic_padding());
```
Comment (joshlf) [inline src/layout.rs]: Remove?

File: src/util/macro_util.rs [hunk 7]
```diff
@@ -359,6 +359,31 @@ macro_rules! struct_has_padding {
...
 }
 
+/// Does the struct type `$t` have padding?
+///
+/// `$ts` is the list of the type of every field in `$t`. `$t` must be a
+/// `repr(C)` struct type, or else `struct_has_padding!`'s result may be
+/// meaningless.
+#[doc(hidden)] // `#[macro_export]` bypasses this module's `#[doc(hidden)]`.
+#[macro_export]
+macro_rules! cstruct_has_padding {
```
Comment (joshlf) [inline src/util/macro_util.rs]:
```suggestion
macro_rules! repr_c_struct_has_padding {
```

I found `cstruct` to be confusing until I saw `repr(C)` in the doc comment.

File: src/util/macro_util.rs [hunk 8]
```diff
@@ -359,6 +359,31 @@ macro_rules! struct_has_padding {
     };
 }
 
+/// Does the struct type `$t` have padding?
```
Comment (joshlf) [inline src/util/macro_util.rs]:
```suggestion
/// Does the `repr(C)` struct type `$t` have padding?
```

File: zerocopy-derive/src/lib.rs [hunk 9]
```diff
@@ -1288,6 +1289,8 @@ fn derive_into_bytes_struct(
         // check for in this branch's condition.
         (None, false)
     } else if ast.generics.params.is_empty() {
+        let is_syntactic_dst =
+            strct.fields().last().map(|(_, _, ty)| matches!(ty, Type::Slice(_))).unwrap_or(false);
```
Comment (joshlf) [inline zerocopy-derive/src/lib.rs:1294]: Can you add a comment explaining this line? IME this won't be easy to read in the future when we've forgotten the context.

File: zerocopy-derive/src/lib.rs [hunk 10]
```diff
@@ -1475,6 +1482,8 @@ enum PaddingCheck {
     /// Check that the sum of the fields' sizes exactly equals the struct's
     /// size.
     Struct,
+    /// Check that a `repr(C)` struct has no padding.
+    CStruct,
```
Comment (joshlf) [inline zerocopy-derive/src/lib.rs]:
```suggestion
    ReprCStruct,
```

Same justification as the macro name

**Review by joshlf — CHANGES_REQUESTED** [review]
Wooo!!!

Can you add UI tests that succeed but wouldn't have succeeded before this change? We may also want some new failure ones to make sure we're properly exercising the edge cases of our logic (namely, what if you have a non-syntactically-DST-but-actually-DST `repr(C)` type? What about a syntactically-DST non-`repr(C)` type? etc).

**Comment by jswrenn** [comment]
> Can you add UI tests that succeed but wouldn't have succeeded before this change? We may also want some new failure ones to make sure we're properly exercising the edge cases of our logic (namely, what if you have a non-syntactically-DST-but-actually-DST `repr(C)` type? What about a syntactically-DST non-`repr(C)` type? etc).

[Done](https://github.com/google/zerocopy/compare/30192700eb00f4f13aeff12d9b48e135e8ff4295..07fd645369aa3b5dff4c73b64890c9c4176f479f)

File: src/layout.rs (hunk 5 above)
Comment (jswrenn) [reply src/layout.rs]: For the two lines in the form here, I'm in favor of this change ([done](https://github.com/google/zerocopy/compare/07fd645369aa3b5dff4c73b64890c9c4176f479f..a692413d419371dccd6313bc0575f525066b31e3#diff-7937a85ae16def66de428ea5dc339277fc8ca7bb7e40c247dc0b485acc39da6cR934-R1002)), but we have eight instances that look basically exactly like this:
```rust
let expected = DstLayout::new_zst(None)
            .extend(DstLayout::for_type::<NotKnownLayout<AU32>>(), None)
            .1
            .extend(<[u8] as KnownLayout>::LAYOUT, None)
            .1
            .pad_to_align()
            .1;
```
I'm not sure much is gained by expanding each of those out and tripling the number of `let`s. Are you sure you want me to go forward with that change here?

File: src/layout.rs (hunk 2 above)
Comment (jswrenn) [reply src/layout.rs]: [Done](https://github.com/google/zerocopy/compare/07fd645369aa3b5dff4c73b64890c9c4176f479f..a692413d419371dccd6313bc0575f525066b31e3#diff-7937a85ae16def66de428ea5dc339277fc8ca7bb7e40c247dc0b485acc39da6cR218)

File: src/util/macro_util.rs (hunk 8 above)
Comment (jswrenn) [reply src/util/macro_util.rs]: [Done](https://github.com/google/zerocopy/compare/07fd645369aa3b5dff4c73b64890c9c4176f479f..a692413d419371dccd6313bc0575f525066b31e3#diff-2abf6454394ca2a0779f4bf7cc0631f23cf1dc484566b5937e8a1ad9a5b9e3e8R342-R362)

File: src/util/macro_util.rs (hunk 7 above)
Comment (jswrenn) [reply src/util/macro_util.rs]: [Done](https://github.com/google/zerocopy/compare/07fd645369aa3b5dff4c73b64890c9c4176f479f..a692413d419371dccd6313bc0575f525066b31e3#diff-2abf6454394ca2a0779f4bf7cc0631f23cf1dc484566b5937e8a1ad9a5b9e3e8R369)

File: zerocopy-derive/src/lib.rs (hunk 9 above)
Comment (jswrenn) [reply zerocopy-derive/src/lib.rs:1294]: [Done](https://github.com/google/zerocopy/compare/07fd645369aa3b5dff4c73b64890c9c4176f479f..a692413d419371dccd6313bc0575f525066b31e3#diff-977c29e6cf079bd7307f811df24f99633df603962785c583145eb6250b0e12daR1292-R1294)

File: src/layout.rs (hunk 6 above)
Comment (jswrenn) [reply src/layout.rs]: [Done](https://github.com/google/zerocopy/compare/07fd645369aa3b5dff4c73b64890c9c4176f479f..a692413d419371dccd6313bc0575f525066b31e3#diff-7937a85ae16def66de428ea5dc339277fc8ca7bb7e40c247dc0b485acc39da6cL1802-L1803)

File: src/layout.rs (hunk 3 above)
Comment (jswrenn) [reply src/layout.rs]: What's the typo here?

Comment (joshlf) [reply src/layout.rs]: Sorry, I meant - can we make this an enum so its meaning is clearer at the call site?

File: src/layout.rs (hunk 4 above)
Comment (joshlf) [reply src/layout.rs]: Clarifying: Can we make this an enum so its meaning is clearer at the call site?

File: src/layout.rs (hunk 5 above)
Comment (joshlf) [reply src/layout.rs]: Agreed; let's keep those as-is.

File: src/layout.rs (hunk 1 above)
Comment (jswrenn) [reply src/layout.rs]: We do boolean operations on the 'is statically padded' component in both implementation (the first two snippets) and call-site (the final snippet):

- https://github.com/google/zerocopy/blob/a692413d419371dccd6313bc0575f525066b31e3/src/layout.rs#L241
- https://github.com/google/zerocopy/blob/a692413d419371dccd6313bc0575f525066b31e3/src/layout.rs#L260
- https://github.com/google/zerocopy/blob/a692413d419371dccd6313bc0575f525066b31e3/src/util/macro_util.rs#L376

I think these sites would be made less clear using a bespoke enum, since we'd need to compensate for the lack of `|=` and `||`. We could add `BitOr` and `Or` implementations, but that would require keeping a mental mapping of which variant of our bespoke enum was truthy and which is falsey.

I'm in favor of bespoke enums in argument position or in a tuple return position that would otherwise carry multiple tuples, but here — as the sole boolean produced by this method (and used in a _very_ boolean fashion) — I think just using a `bool` is clearer.

Comment (jswrenn) [reply src/layout.rs]: I know there's something of an anti-bool sentiment in engineering right now, but the two mostly widely-read anti-bool manifestos cite argument-position bools (not return position bools):
- https://www.luu.io/posts/2019-dont-use-booleans
- https://alexkondov.com/should-you-pass-boolean-to-functions/

Comment (joshlf) [reply src/layout.rs]: My concern is that the function names don't make it clear what the meaning of the `bool` is (`requires_dynamic_padding` is an obvious exception, and I'm fine with that returning a `bool`).

If we returned an enum, we could do a few things to make the usage site more concise:
- Implement `Eq`, permitting e.g. `static_padding |= (interfield_padding == Static)`
- Add a method, permitting e.g. `static_padding |= interfield_padding.is_static()`
- Probably others I'm not thinking of

Comment (joshlf) [reply src/layout.rs]: Obsoleted by recent changes.

File: src/layout.rs (hunk 3 above)
Comment (joshlf) [reply src/layout.rs]: Obsoleted by recent changes.

File: src/layout.rs (hunk 4 above)
Comment (joshlf) [reply src/layout.rs]: Obsoleted by recent changes.
